src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs

```diff
@@ -11,60 +11,72 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
 {
     internal sealed partial class VisualStudioRuleSetManager
     {
-        private sealed class RuleSetFile : IRuleSetFile
+        private sealed class RuleSetFile : IRuleSetFile, IDisposable
         {
-            private readonly string _filePath;
-            private readonly List<FileChangeTracker> _trackers;
             private readonly VisualStudioRuleSetManager _ruleSetManager;
+            private readonly object _gate = new object();
+
+            /// <summary>
+            /// The list of file trackers we have. This is null if we haven't even computed trackers yet; but can be empty if we
+            /// already know the file was modified and we unsubscribed.
+            /// </summary>
+            private List<FileChangeTracker> _trackers;
 
             private ReportDiagnostic _generalDiagnosticOption;
             private ImmutableDictionary<string, ReportDiagnostic> _specificDiagnosticOptions;
             private bool _subscribed = false;
             private bool _optionsRead = false;
+            private bool _removedFromRuleSetManager = false;
 
             private Exception _exception;
 
-            public RuleSetFile(string filePath, IVsFileChangeEx fileChangeService, VisualStudioRuleSetManager ruleSetManager)
+            public RuleSetFile(string filePath, VisualStudioRuleSetManager ruleSetManager)
             {
-                _filePath = filePath;
+                FilePath = filePath;
                 _ruleSetManager = ruleSetManager;
+            }
 
-                ImmutableArray<string> includes;
-
-                try
-                {
-                    includes = RuleSet.GetEffectiveIncludesFromFile(filePath);
-                }
-                catch (Exception e)
+            public void InitializeFileTracking(IVsFileChangeEx fileChangeService)
```

Comment:
:bulb: `internal` so it's clear on review that it can't be part of `IRuleSetFile`

Reply:
Won't fix: HeeJae asked why it was internal. At this point not going to play flip flop.